### PR TITLE
Add limit for matchfuzzy()

### PR DESCRIPTION
**Pull request by mattn**
matchfuzzy() is useful. But when huge list are given, it takes heavily times. This pull-request add limit for matchfuzzy().

### CODE REVIEW

File: src/testdir/test_matchfuzzy.vim [hunk 1]
```diff
@@ -230,4 +230,13 @@ func Test_matchfuzzypos_mbyte()
   call assert_equal([['xффйд'], [[2, 3, 4]], [168]], matchfuzzypos(['xффйд'], 'фйд'))
 endfunc
 
+" Test for matchfuzz() with limit
```
Comment (zeertzjq) [inline src/testdir/test_matchfuzzy.vim]:
```suggestion
" Test for matchfuzzy() with limit
```

**Comment by LemonBoy** [comment]
Since there's already a trailing `{dict}` argument why not make this limit part of the option dictionary?

**Comment by mattn** [comment]
Ah, you are right. I'll fix soon.

**Comment by mattn** [comment]
done

File: src/search.c [hunk 2]
```diff
@@ -4879,6 +4888,17 @@ do_fuzzymatch(typval_T *argvars, typval_T *rettv, int retmatchpos)
...
 	    }
 	}
+	else if ((di = dict_find(d, (char_u *)"limit", -1)) != NULL)
+	{
+	    if (di->di_tv.v_type != VAR_NUMBER)
+	    {
+		semsg(_(e_invalid_argument), tv_get_string(&di->di_tv));
+		return;
+	    }
+	    key = tv_get_string(&di->di_tv);
```
Comment (mattn) [inline src/search.c:4898]: wrong

File: src/search.c [hunk 3]
```diff
@@ -4648,13 +4648,14 @@ fuzzy_match_in_list(
 	char_u		*key,
 	callback_T	*item_cb,
 	int		retmatchpos,
-	list_T		*fmatchlist)
+	list_T		*fmatchlist,
+	long		num_limit)
```
Comment (yegappan) [inline src/search.c]: Can you rename 'num_limit' to 'max_matches'?

File: src/search.c [hunk 4]
```diff
@@ -4675,6 +4676,13 @@ fuzzy_match_in_list(
 	ptrs[i].idx = i;
 	ptrs[i].item = li;
 	ptrs[i].score = SCORE_NONE;
+
+	if (num_limit > 0 && found_match >= num_limit)
```
Comment (yegappan) [inline src/search.c]: One issue with placing a limit on the maximum number of matches is that the best match may be found after this limit. 
 In that case, matchfuzzy() will return the best match within this limit.

File: src/search.c [hunk 5]
```diff
@@ -4822,6 +4830,7 @@ do_fuzzymatch(typval_T *argvars, typval_T *rettv, int retmatchpos)
     char_u	*key = NULL;
     int		ret;
     int		matchseq = FALSE;
+    long	num_limit = 0;
```
Comment (yegappan) [inline src/search.c]: Can you rename this to 'max_matches'?

File: runtime/doc/builtin.txt [hunk 6]
```diff
@@ -5589,6 +5589,7 @@ matchfuzzy({list}, {str} [, {dict}])			*matchfuzzy()*
 				This should accept a dictionary item as the
 				argument and return the text for that item to
 				use for fuzzy matching.
+		    limit	the maximum number of list to be matched.
```
Comment (yegappan) [inline runtime/doc/builtin.txt]: Can you change this to "maximum number of items in {list} to be matched"?

File: src/testdir/test_matchfuzzy.vim [hunk 7]
```diff
@@ -230,4 +230,15 @@ func Test_matchfuzzypos_mbyte()
...
 
+" Test for matchfuzzy() with limit
+func Test_matchfuzzy_limit()
+  let x = ['1', '2', '3', '2']
+  call assert_equal(['2', '2'], x->matchfuzzy('2'))
+  call assert_equal(['2', '2'], x->matchfuzzy('2', #{}))
+  call assert_equal(['2', '2'], x->matchfuzzy('2', #{limit: 0}))
+  call assert_equal(['2'], x->matchfuzzy('2', #{limit: 1}))
+  call assert_equal(['2', '2'], x->matchfuzzy('2', #{limit: 2}))
+  call assert_equal(['2', '2'], x->matchfuzzy('2', #{limit: 3}))
```
Comment (yegappan) [inline src/testdir/test_matchfuzzy.vim:241]: Can you also add a test for passing a non-numeric value to 'limit'?

**Comment by brammool** [comment]
Looking at the code I see that the sorting happens before dropping items that don't match.  It would be more efficient to only put matches in ptrs[].  Then when a limit is given it only needs that many entries.
This can be done in a followup PR.

**Pull request closed without merging**